spec/client/ics-012-near-client/README.md

```diff
@@ -0,0 +1,550 @@
+
+```typescript
+type PublicKey = ED25519PublicKey
+type Signature = ED25519Signature
+```
+
+`borsh` is a generic serialization function which follows the [Borsh serialization format](https://borsh.io/).
+
+`merklize` is a generic function which can construct a merkle tree from an array which the element in it can be serialized by `borsh`. This function should return the merkle root of the tree at least.
+
+In NEAR protocol, the block producers are changed by time (about 12 hours). The period is known as `epoch` and the id of an epoch is represented by a `CryptoHash`.
+
+### Desired Properties
+
+This specification must satisfy the client interface defined in ICS 2.
+
+## Technical Specification
+
+This specification is based on the [NEAR light client specification](https://nomicon.io/ChainSpec/LightClient) and the implementation of [nearcore v1.30.0](https://github.com/near/nearcore/releases/tag/1.30.0) by adding necessary data fields and checking processes.
+
+### Client state
+
+The NEAR client state tracks the following data:
+
+```typescript
+interface ClientState {
+    trustingPeriod: uint64
+    latestHeight: Height
+    latestTimestamp: uint64
+    frozenHeight: Maybe<uint64>
+    upgradeCommitmentPrefix: []byte
+    upgradeKey: []bype
+}
+```
+
+### Consensus state
+
+The NEAR client tracks the block producers of current epoch and header (refer to [Headers section](#headers)) for all previously verified consensus states (these can be pruned after a certain period, but should not be pruned beforehand).
+
+```typescript
+interface ValidatorStakeView {
+    accountId: String
+    publicKey: PublicKey
+    stake: uint128
+}
+
+interface ConsensusState {
+    currentBps: List<ValidatorStakeView>
+    header: Header
+}
+```
+
+### Height
+
+The height of a NEAR client is an `uint64` number.
+
+```typescript
+interface Height {
+    height: uint64
+}
+```
+
+Comparison between heights is implemented as follows:
+
+```typescript
+function compare(a: Height, b: Height): Ord {
+    if (a.height < b.height)
+        return LT
+    else if (a.height === b.height)
+        return EQ
+    return GT
+}
+```
+
+### Headers
+
+The NEAR client headers include the `LightClientBlock` and previou state root of chunks. The entire block producers for next epoch and approvals for the block after the next are included in `LightClientBlock`.
+
+```typescript
+interface BlockHeaderInnerLite {
+    height: Height
+    epochId: CryptoHash
+    nextEpochId: CryptoHash
+    prevStateRoot: CryptoHash
+    outcomeRoot: CryptoHash
+    timestamp: uint64   // in nanoseconds
+    nextBpHash: CryptoHash
+    blockMerkleRoot: CryptoHash
+}
+
+interface LightClientBlock {
+    prevBlockHash: CryptoHash
+    nextBlockInnerHash: CryptoHash,
+    innerLite: BlockHeaderInnerLite
+    innerRestHash: CryptoHash
+    nextBps: Maybe<List<ValidatorStakeView>>
+    approvalsAfterNext: List<Maybe<Signature>>
```

Comment:
As in function `verifyHeader` we also need stake amount of validators to check approved stake amount (step 4 and 5 in this function), I think the list here should be `List<ValidatorStake>` rather than `List<Signature>`.
And the `Maybe` for this field is also better, because this field could be `null` while retrieving from RPC interface of NEAR protocol.
I prefer keep current design. But of course, the `ValidatorStakeView` should be renamed to `ValidatorStake`.

Reply:
Sorry for the long time that has taken me to reply... Maybe we are talking about two different things... I meant that `approvalsAfterNext: List<Maybe<Signature>>` could be `approvalsAfterNext: List<Signature>` since in `verifyHeader` there's this code to check for `null` in line 289:

```
if maybeSignature === null {
  continue
}
```         

And then in line 297 is passed to `verifySignature`:

```        
assert(verifySignature(
  public_key: blockProducer.public_key,
  signature: maybeSignature,
  message: approvalMessage
))
```

But looks like the `Maybe` type doesn't really provide any extra benefit.

Reply:
I got your point now. Here's an example to explain the case:
We have 30 block producers in `nextBps`. But the signatures don't need to be 30. If the producer with index 20 doesn't sign the message, then the signature with index 20 in `approvalsAfterNext` should be `null`, to ensure the orders of signatures are exactly match with the producers in `nextBps`.
This is why the `approvalsAfterNext` should be defined as `List<Maybe<Signature>>` rather than `List<Signature>`.
Besides, the data definitions are based on the real data which can be queried by NEAR rpc interfaces. I think there is no need to change them in the spec.